### PR TITLE
Quest entrance check refactoring

### PR DESCRIPTION
**Pull request by julealgon**
Replaces some low-level X/Y offset checks for quest entrance check with some higher level abstractions.

### CODE REVIEW

**Comment by julealgon** [comment]
Depends on https://github.com/diasurgical/devilutionX/pull/2253 being merged first.

**Review by AJenbo — DISMISSED** [review]

File: Source/quests.cpp [hunk 1]
```diff
@@ -885,4 +873,14 @@ bool QuestStruct::IsAvailable()
 	return true;
 }
 
+bool QuestStruct::EntranceBoundaryContains(Point position) const
```
Comment (ephphatha) [inline Source/quests.cpp:876]: I think this is the same shape as used for town portal portals, potentially useful to pull this out of the QuestStruct class.

**Comment by AJenbo** [comment]
```
      Start 66: QuestTest.SinglePlayerBadPools
66/98 Test #66: QuestTest.SinglePlayerBadPools ......................***Exception: SegFault  0.02 sec
Note: Google Test filter = QuestTest.SinglePlayerBadPools
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from QuestTest
[ RUN      ] QuestTest.SinglePlayerBadPools

```

**Comment by julealgon** [comment]
> ```
>       Start 66: QuestTest.SinglePlayerBadPools
> 66/98 Test #66: QuestTest.SinglePlayerBadPools ......................***Exception: SegFault  0.02 sec
> Note: Google Test filter = QuestTest.SinglePlayerBadPools
> [==========] Running 1 test from 1 test suite.
> [----------] Global test environment set-up.
> [----------] 1 test from QuestTest
> [ RUN      ] QuestTest.SinglePlayerBadPools
> ```

I have no idea what is happening there. Any hints?

**Comment by AJenbo** [comment]
The change is causing one of the RNG tests to crash the application as best I can tell. You would probably need to run the test locally to figure out exactly what is triggering it.

**Comment by obligaron** [comment]
I run the test locally.
The generated random value of this [line](https://github.com/diasurgical/devilutionX/blob/60e27e2427e4d125b533b0855fb4116ab455209d/Source/quests.cpp#L357) is -2.
So we try to access `QuestGroup1` with a negative index => Undefined behavior. 😭 
This pr changes something that changes the memory layout, so we now don't override some random memory but get a access violation exception.

Is there a way, that we can fix this? So we don't write random memory but still stay compatible? 🤔 

@ephphatha To your notice, cause you wrote the test cases and might be interested. 🙂

**Comment by ephphatha** [comment]
> Is there a way, that we can fix this? So we don't write random memory but still stay compatible? 

I'd been lazy about fixing it with the current rng code (there's a fix in #2261 along with the proposed api).

```cpp
int randomIndex = GenerateRnd(sizeof(QuestGroup1) / sizeof(int));
if (randomIndex >= 0)
	quests[QuestGroup1[randomIndex]]._qactive = QUEST_NOTAVAIL;
randomIndex = GenerateRnd(sizeof(QuestGroup2) / sizeof(int));
if (randomIndex >= 0)
	quests[QuestGroup2[randomIndex]]._qactive = QUEST_NOTAVAIL;
randomIndex = GenerateRnd(sizeof(QuestGroup3) / sizeof(int));
if (randomIndex >= 0)
	quests[QuestGroup3[randomIndex]]._qactive = QUEST_NOTAVAIL;
randomIndex = GenerateRnd(sizeof(QuestGroup4) / sizeof(int));
if (randomIndex >= 0) // always true, QuestGroup4 has two members
	quests[QuestGroup4[randomIndex]]._qactive = QUEST_NOTAVAIL;
```

**Comment by julealgon** [comment]
What are the next steps here? Should I mark this as draft until the fix is made to the RNG logic?

**Comment by AJenbo** [comment]
> What are the next steps here? Should I mark this as draft until the fix is made to the RNG logic?

Either that, or just leave it like this for now.

**Review by AJenbo — APPROVED** [review]

File: Source/quests.cpp (hunk 1 above)
Comment (AJenbo) [reply Source/quests.cpp]: Indeed CheckTown and CheckRportal have the same checks

Comment (AJenbo) [reply Source/quests.cpp]: Plz see https://github.com/diasurgical/devilutionX/pull/2845

**Pull request merged**